### ISSUE TITLE
`crypto_pwhash_argon2id_str_verify` causes signal 11 crash if memory allocation fails in initialize

### ISSUE THREAD

**Issue by cmouse** [jedisct1/libsodium#758]
Test program

```.c
#include <sodium.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <sys/resource.h>

int main(void) {
	const char *pw = "$argon2id$v=19$m=1048576,t=7,p=1$kIhnUR13GrtOvvpbJNJmnQ$o7O6Whxs3s8IE09yY9S2dPkJjJyEVc78GRFilYVS9fU";
	const char *pt = "This is a test";
	struct rlimit lim = {
		.rlim_cur = 1038570*1024,
		.rlim_max = 1038570*1024,
	};
	/* set memory limit */
	if (setrlimit(RLIMIT_AS, &lim) != 0)
		perror("setrlimit(AS): ");
	if (setrlimit(RLIMIT_DATA, &lim) != 0)
		perror("setrlimit(DATA): ");
	if (setrlimit(RLIMIT_STACK, &lim) != 0)
		perror("setrlimit(STACK): ");

	if (sodium_init() < 0) {
		printf("Failed to init\n");
		abort();
	}
	printf("Verifying: ");
	if (crypto_pwhash_str_verify(pw, pt, strlen(pt)) < 0)
		printf("NOK\n");
	else
		printf("OK\n");
	return 0;
}
```

The culprit is in `free_memory` function in argon2-core.c which does not take into account that while `instance->region` is a valid pointer, the `memory` member is NULL pointer.

**Comment by jedisct1** [comment]
Unfortunately, I wasn't able to reproduce this :(

On what operating system does it happen?

**Comment by cmouse** [comment]
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.11 (jessie)
Release:	8.11
Codename:	jessie

Version: libsodium-dev:amd64  1:1.0.15-1

```(gdb) bt
#0  memset () at ../sysdeps/x86_64/memset.S:78
#1  0x00007ffff7ba4e26 in memset (__len=1073741824, __ch=0, __dest=0x0) at /usr/include/x86_64-linux-gnu/bits/string3.h:84
#2  sodium_memzero (pnt=0x0, len=1073741824) at ../../../src/libsodium/sodium/utils.c:90
#3  0x00007ffff7b9d676 in clear_memory (clear=<optimized out>, instance=instance@entry=0x7fffffffe240) at ../../../src/libsodium/crypto_pwhash/argon2/argon2-core.c:139
#4  free_instance (instance=instance@entry=0x7fffffffe240, flags=<optimized out>) at ../../../src/libsodium/crypto_pwhash/argon2/argon2-core.c:174
#5  0x00007ffff7b9dd9d in initialize (instance=instance@entry=0x7fffffffe240, context=context@entry=0x7fffffffe2a0) at ../../../src/libsodium/crypto_pwhash/argon2/argon2-core.c:566
#6  0x00007ffff7b9fccb in argon2_ctx (context=0x7fffffffe2a0, type=Argon2_id) at ../../../src/libsodium/crypto_pwhash/argon2/argon2.c:66
#7  0x00007ffff7b9fde7 in argon2_hash (t_cost=7, m_cost=<optimized out>, parallelism=1, pwd=pwd@entry=0x400a5c, pwdlen=pwdlen@entry=14, salt=0x601080, saltlen=16, hash=0x601160, hashlen=32, encoded=0x0, 
    encodedlen=0, type=Argon2_id) at ../../../src/libsodium/crypto_pwhash/argon2/argon2.c:125
#8  0x00007ffff7ba00b5 in argon2_verify (encoded=0x4009f8 "$argon2id$v=19$m=1048576,t=7,p=1$kIhnUR13GrtOvvpbJNJmnQ$o7O6Whxs3s8IE09yY9S2dPkJjJyEVc78GRFilYVS9fU", pwd=0x400a5c, pwdlen=14, 
    type=type@entry=Argon2_id) at ../../../src/libsodium/crypto_pwhash/argon2/argon2.c:248
#9  0x00007ffff7ba019a in argon2id_verify (encoded=<optimized out>, pwd=<optimized out>, pwdlen=<optimized out>) at ../../../src/libsodium/crypto_pwhash/argon2/argon2.c:272
#10 0x00007ffff7ba0e10 in crypto_pwhash_argon2id_str_verify (str=<optimized out>, passwd=<optimized out>, passwdlen=<optimized out>) at ../../../src/libsodium/crypto_pwhash/argon2/pwhash_argon2id.c:226
#11 0x0000000000400944 in main () at test.c:28
```

**Comment by jedisct1** [comment]
Can you try with the current stable version (https://download.libsodium.org/libsodium/releases/LATEST.tar.gz) or even the base 1.0.16 release (https://download.libsodium.org/libsodium/releases/libsodium-1.0.16.tar.gz)?

**Comment by cmouse** [comment]
I can confirm now that it crashes on 1.0.15, and is fixed on 1.0.16.
